docs: upgrade Keptn to Helm from a manifest installation (#2270)
keptn/lifecycle-toolkit: docs/content/en/docs/install/upgrade.md

additional steps are required to use the Helm Chart to upgrade.
Contact us on Slack for assistance.

## Upgrade to Helm from a manifest installation

> **Warning**

> Upgrade to Helm from a manifest installation can cause data loss.

Reviewer: I would specify what we expect to lose: CR will be lost in the upgrade process

> Upgrade to Helm from a manifest installation can cause data loss.

Author: ```suggestion
Upgrade to Helm from a manifest installation can cause loss
of Keptn custom resources.
```

additional steps are required to use the Helm Chart to upgrade.
Contact us on Slack for assistance.

> ## Upgrade from installation via manifests using Helm

Reviewer 2: ```suggestion
## Upgrade to Helm from a manifest installation
```

Keptn v.0.7.0 and later can be installed with Helm charts;
Keptn v.0.8.3 can only be installed with Helm.
If you previously installed Keptn from manifests,

> you can not directly upgrade to with Helm but must back up your Keptn data,

Reviewer: ```suggestion
you can not directly upgrade with Helm but must back up your Keptn data,
```

> If you are trying to upgrade Keptn installed via manifests to a version
> which does not anymore support manifests installation (>= 0.8.3) or any
> other version which supports installation via Helm (>= 0.7.0) you should
> be aware that there might be a certain loose of data.
> 
> Unfortunatelly, there is no straight way how to upgrade Keptn installed via
> manifests to a higher version via Helm.
> There is a need to uninstall Keptn
> and install it afterwards via Helm.

Reviewer 2: ```suggestion
Keptn v.0.7.0 and later can be installed with Helm charts;
Keptn v.0.8.3 can only be installed with Helm.
If you previously installed Keptn from manifests,
you can not directly upgrade to with Helm but must back up your Keptn data,
then reinstall Keptn from a Helm chart
and you may lose some data.
```

> To not loose all of your data, we encourage you to do a backup of the Keptn CRs,
> `Namespaces`, `Secrets` and `ConfigMaps`.

Reviewer 2: ```suggestion
1. Back up all the Keptn CRs,
`Namespaces`, `Secrets` and `ConfigMaps`.
```

> To create a backup, copy the following code into `backup-script.sh` file:

Reviewer 2: ```suggestion
To do this, copy the following code into `backup-script.sh` file:
```

```shell
#!/bin/bash

kubectl get ns -oyaml >> backup.yaml
kubectl get configmaps -A -oyaml >> backup.yaml
kubectl get secrets -A -oyaml >> backup.yaml

resources=$(kubectl get crd -n keptn-lifecycle-toolkit-system --no-headers -o custom-columns=NAME:.metadata.name)

for name in "${resources[@]}"; do
   kubectl get $name -A -oyaml >> backup.yaml
done
```

> next enable execution bit of your file and execute the script:

Reviewer 2: ```suggestion
Then make your file executable and execute the script:
```

```shell
chmod +x backup-script.sh && ./backup-script.sh
```

> This will create a manifest `backup.yaml` with all the CRs, `Namespaces`, `Secrets`

Reviewer 2: ```suggestion
This creates a manifest called `backup.yaml` that includes all the CRs, `Namespaces`, `Secrets`
```

and `ConfigMaps`.

> > **Note** Please be aware that this is not a backup of your whole cluster and this part
> you should hadle by yourself.

Reviewer 2: ```suggestion
> **Note** This is only backs up you Keptn configuration data.
   Use standard Kuberetes tools and practices to back up your entire cluster.
```

> To proceed with the upgrade, you need to completely remove your Keptn installation:

Reviewer 2: ```suggestion
1. Completely remove your Keptn installation with the following command sequence:
```

```shell
your-keptn-version=<your-keptn-version>

> kubectl delete -f https://github.com/keptn/lifecycle-toolkit/releases/download/$your-keptn-version/manifest.yaml

Reviewer 2: ```suggestion
kubectl delete -f \
     https://github.com/keptn/lifecycle-toolkit/releases/download/$your-keptn-version/manifest.yaml
```

```

> and create a new clean installation of Keptn via Helm:

Reviewer 2: ```suggestion
1.  Use Helm Keptn:
```

```shell
helm repo add klt https://charts.lifecycle.keptn.sh
helm repo update
helm upgrade --install keptn klt/klt -n keptn-lifecycle-toolkit-system --create-namespace --wait
```

> To check advanced installation options, refer to the [installation section](install.md).

Reviewer 2: ```suggestion
To check advanced installation options, refer to
[Modify Helm configuration options](install.md/#modify-helm-configuration-options).
```

> After finishing your installation, your system is clean and prepared to accept the
> resources from your backup.

Reviewer 2: ```suggestion
After the installation finishes, restire the Keptn resources from your backup:
```

To apply them execute:

```shell
kubectl apply -f backup.yaml
```

> **Note** Please be aware that some Keptn applications may start the deployment from the
beginning (if the `Deployments` are in place) and the system is not guaranteed to return

> to the exact state it was in before re-installation, even if you created the backup correctly.

Reviewer: can you apply a CR with a Stuatus? Won't that field be lost?

Author: you can and yes it will be lost, but that is expected. The backup is not 100% and we declare that the whole process will start from scratch if the resources would be applied, we also declare in the docs that there is no guarantee. Users need to count with loose of data

Reviewer 3: what's the purpose of the backup then?! 🤔

Author: docs adapted according to our private discussion to not create a backup of all Keptn resources but only the manifests which were applied to the cluster by user (or GitOps solution)

> > **Note** Please be aware that all Keptn applications will start the deployment from the start and
> there is no guerantee that the system will return to the state before re-installation, eve if you
> created the backup correctly.

Reviewer 2: ```suggestion
> **Note** Please be aware that some Keptn applications may not start the deployment from the beginning and
the system is not guaranteed to return to the exact state it was in before re-installation,
even if you created the backup correctly.
```
